components/playlist/renderer/playlist_render_frame_observer.cc

```diff
@@ -35,137 +66,108 @@ void PlaylistRenderFrameObserver::OnDestruct() {
   delete this;
 }
 
-bool PlaylistRenderFrameObserver::EnsureConnectedToMediaHandler() {
-  if (!media_handler_.is_bound()) {
+void PlaylistRenderFrameObserver::AddMediaSourceAPISuppressor(
+    const std::string& media_source_api_suppressor) {
+  DVLOG(2) << __FUNCTION__;
+
+  media_source_api_suppressor_ = media_source_api_suppressor;
+  CHECK(!media_source_api_suppressor_->empty());
+}
+
+void PlaylistRenderFrameObserver::AddMediaDetector(
+    const std::string& media_detector) {
+  DVLOG(2) << __FUNCTION__;
+
+  media_detector_ = media_detector;
+  CHECK(!media_detector_->empty());
+}
+
+void PlaylistRenderFrameObserver::SetUpForTesting() {
+  testing_ = true;
+}
+
+void PlaylistRenderFrameObserver::BindConfigurator(
+    mojo::PendingAssociatedReceiver<
+        mojom::PlaylistRenderFrameObserverConfigurator> receiver) {
+  configurator_receiver_.reset();
+  configurator_receiver_.Bind(std::move(receiver));
+}
+
+const mojo::Remote<playlist::mojom::PlaylistMediaHandler>&
+PlaylistRenderFrameObserver::GetMediaHandler() {
+  if (!media_handler_) {
     render_frame()->GetBrowserInterfaceBroker()->GetInterface(
         media_handler_.BindNewPipeAndPassReceiver());
-    media_handler_.set_disconnect_handler(
-        base::BindOnce(&PlaylistRenderFrameObserver::OnMediaHandlerDisconnect,
-                       weak_ptr_factory_.GetWeakPtr()));
+    media_handler_.reset_on_disconnect();
   }
 
-  return media_handler_.is_bound();
-}
-
-void PlaylistRenderFrameObserver::OnMediaHandlerDisconnect() {
-  media_handler_.reset();
-  EnsureConnectedToMediaHandler();
+  return media_handler_;
 }
 
 void PlaylistRenderFrameObserver::RunScriptsAtDocumentStart() {
-  if (render_frame()->GetWebFrame()->IsProvisional()) {
-    return;
-  }
+  if (media_source_api_suppressor_) {
+    v8::Isolate* isolate = blink::MainThreadIsolate();
+    v8::Isolate::Scope isolate_scope(isolate);
+    v8::HandleScope handle_scope(isolate);
 
-  const auto& blink_preferences = render_frame()->GetBlinkPreferences();
-  if (blink_preferences.hide_media_src_api) {
-    HideMediaSourceAPI();
+    Inject(*media_source_api_suppressor_,
+           render_frame()->GetWebFrame()->MainWorldScriptContext());
   }
+}
 
-  if (blink_preferences.should_detect_media_files) {
-    InstallMediaDetector();
+void PlaylistRenderFrameObserver::RunScriptsAtDocumentEnd() {
+  if (media_detector_) {
+    v8::Isolate* isolate = blink::MainThreadIsolate();
+    v8::Isolate::Scope isolate_scope(isolate);
+    v8::HandleScope handle_scope(isolate);
+
+    v8::Local<v8::Context> context =
+#if !BUILDFLAG(IS_ANDROID)
+        render_frame()->GetWebFrame()->GetScriptContextFromWorldId(
+            isolate, isolated_world_id_);
+#else
+        render_frame()->GetWebFrame()->MainWorldScriptContext();
```

Comment:
Why are we running this in the main world? Didn't it previously execute in an isolated world? Running it in the main world is potentially dangerous because of prototype pollution

Reply:
There's no change in behavior (yet): Android injects into main, desktop injects into `ISOLATED_WORLD_ID_BRAVE_INTERNAL`. IIRC, that's because Android needs https://github.com/brave/playlist-component/blob/main/youtube.com.js to extract thumbnails (which uses `window.ytplayer...`, hence needs main world access), as the necessary OG `<meta>` tags (which is how desktop does it) are not present on Android.

Reply:
I believe you have already filed an issue for this?

Reply:
Not yet — raising now.

Reply:
https://github.com/brave/brave-browser/issues/36443